tools/storm/e2e/discover.go

```diff
@@ -1,182 +1,134 @@
+func DiscoverTridentScenarios(log *logrus.Logger) ([]scenario.TridentE2EScenario, error) {
+	rawConfigs, err := content.ReadFile("configurations/configurations.yaml")
 	if err != nil {
-		log.Errorf("Failed to read configurations directory: %v", err)
-		return nil
+		return nil, fmt.Errorf("failed to read configurations file: %w", err)
 	}
 
-	var tridentConfigs = make(map[string]*tridentConfig)
+	var cfg configs
+	err = yaml.Unmarshal(rawConfigs, &cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse e2e configurations file: %w", err)
+	}
 
-	for _, entry := range entries {
-		if !entry.IsDir() {
-			continue
-		}
+	var tridentE2EScenarios []scenario.TridentE2EScenario = make([]scenario.TridentE2EScenario, 0, len(cfg))
 
-		configPath := "configurations/trident_configurations/" + entry.Name() + "/trident-config.yaml"
+	for name, conf := range cfg {
+		configPath := getConfigPath(name)
 
 		configYaml, err := content.ReadFile(configPath)
 		if err != nil {
 			log.Fatalf("Failed to read configuration file: %v", err)
 		}
 
-		scenario := CreateTridentScenario(entry.Name())
+		var hostConfig map[string]any
+		err = yaml.Unmarshal(configYaml, &hostConfig)
+		if err != nil {
+			log.Fatalf("Failed to unmarshal configuration file for '%s': %v", name, err)
+		}
 
-		err = yaml.Unmarshal(configYaml, &scenario.config)
+		scenarios, err := produceScenariosFromConfig(name, conf, hostConfig)
 		if err != nil {
-			log.Fatalf("Failed to unmarshal configuration file for '%s': %v", scenario.Name(), err)
+			log.Fatalf("Failed to produce scenarios for '%s': %v", name, err)
 		}
 
-		tridentConfigs[entry.Name()] = &tridentConfig{scenario: scenario, used: false}
+		tridentE2EScenarios = append(tridentE2EScenarios, scenarios...)
 	}
 
-	// Check that all targets exist and that all configs have at least one target
-	for stagePath, targets := range STAGE_PATHS_TARGETS {
-		for _, target := range targets {
-			config, ok := tridentConfigs[target]
-			if !ok {
-				log.Fatalf("Target '%s' in stage path '%s' references non-existing configuration", target, stagePath)
-			}
-			config.used = true
-			config.scenario.AddStagePath(stagePath.String())
-		}
+	return tridentE2EScenarios, nil
+}
+
+func getConfigPath(scenarioName string) string {
+	return "configurations/trident_configurations/" + scenarioName + "/trident-config.yaml"
+}
+
+func produceScenariosFromConfig(name string, conf scenarioConfig, hostConfig map[string]interface{}) ([]scenario.TridentE2EScenario, error) {
+	var scenarios []scenario.TridentE2EScenario
+
+	groups := []struct {
+		hardware scenario.HardwareType
+		runtime  scenario.RuntimeType
+		ring     testrings.TestRing
+	}{
+		{scenario.HardwareTypeBM, scenario.RuntimeTypeHost, conf.Bm.Host},
+		{scenario.HardwareTypeBM, scenario.RuntimeTypeContainer, conf.Bm.Container},
+		{scenario.HardwareTypeVM, scenario.RuntimeTypeHost, conf.Vm.Host},
+		{scenario.HardwareTypeVM, scenario.RuntimeTypeContainer, conf.Vm.Container},
```

Comment:
nit: can the ring host vs container be derived from each other somehow rather than specifying both?